docs/source/onnx.rst

```diff
@@ -994,259 +673,72 @@ you can create and register your own custom ops implementation in PyTorch. Here'
+  to do so, you will need to manually specify the datatype. This often happens with
+  scripted models, where the datatypes are not recorded. For example::
 
     class ImplicitCastType(torch.jit.ScriptModule):
         @torch.jit.script_method
         def forward(self, x):
-            # Exporter knows x is float32, will export '2' as float32 as well.
+            # Exporter knows x is float32, will export "2" as float32 as well.
             y = x + 2
-            # Without type propagation, exporter doesn't know the datatype of y.
-            # Thus '3' is exported as int64 by default.
+            # Currently the exporter doesn't know the datatype of y, so
+            # "3" is exported as int64, which is wrong!
             return y + 3
-            # The following will export correctly.
+            # To fix, replace the line above with:
             # return y + torch.tensor([3], dtype=torch.float32)
 
     x = torch.tensor([1.0], dtype=torch.float32)
-    torch.onnx.export(ImplicitCastType(), x, 'models/implicit_cast.onnx',
+    torch.onnx.export(ImplicitCastType(), x, "implicit_cast.onnx.pb",
                       example_outputs=ImplicitCastType()(x))
 
-Q: Is tensor in-place indexed assignment like `data[index] = new_data` supported?
-
-  Yes, this is supported for ONNX opset version >= 11. Please checkout `Indexing`_.
+  We are trying to improve the datatype propagation in the exporter such that implicit casting
+  is supported in more cases.
 
-Q: Is tensor list exportable to ONNX?
+Q: Are lists of Tensors exportable to ONNX?
 
-  Yes, this is supported now for ONNX opset version >= 11. ONNX introduced the concept of Sequence in opset 11.
-  Similar to list, Sequence is a data type that contains arbitrary number of Tensors.
-  Associated operators are also introduced in ONNX, such as SequenceInsert, SequenceAt, etc.
-  However, in-place list append within loops is not exportable to ONNX. To implement this, please use inplace
-  add operator.
-  E.g.: ::
-
-    class ListLoopModel(torch.nn.Module):
-        def forward(self, x):
-            res = []
-            res1 = []
-            arr = x.split(2, 0)
-            res2 = torch.zeros(3, 4, dtype=torch.long)
-            for i in range(len(arr)):
-                res += [arr[i].sum(0, False)]
-                res1 += [arr[-1 - i].sum(0, False)]
-                res2 += 1
-            return torch.stack(res), torch.stack(res1), res2
-
-    model = torch.jit.script(ListLoopModel())
-    inputs = torch.randn(16)
-
-    out = model(inputs)
-    torch.onnx.export(model, (inputs, ), 'loop_and_list.onnx', opset_version=11, example_outputs=out)
-
-    # onnxruntime
-    import onnxruntime
-    sess = onnxruntime.InferenceSession('loop_and_list.onnx')
-    out_ort = sess.run(None, {
-        sess.get_inputs()[0].name: inputs.numpy(),
-    })
-
-    assert [torch.allclose(o, torch.tensor(o_ort)) for o, o_ort in zip(out, out_ort)]
-
-Use external data format
```

Comment:
Can we keep the documentation about Training and use_external_data_format?
Do you think this info is redundant or not useful?

Reply:
It's redundant with the documentation for the arguments of `torch.onnx.export()`.
When this page is rendered into HTML, that documentation appears on the same page:
https://garymm.github.io/onnx.html